Support virtual (not persisted) generated columns on PostgreSQL 18+

yahonda opened the pull request:
### Motivation / Background

This commit adds support for virtual generated columns on PostgreSQL 18 and later.

```ruby
create_table :users do |t|
    t.string :name
    t.virtual :lower_name,  type: :string,  as: "LOWER(name)", stored: false
    t.virtual :name_length, type: :integer, as: "LENGTH(name)"
end
```

### Detail

PostgreSQL 18 Beta 1 introduces virtual (not persisted) generated columns, which are now the default unless the STORED option is explicitly specified. Rails previously supported only stored generated columns on PostgreSQL 12+, via https://github.com/rails/rails/pull/41856, as only stored generated columns were supported by PostgreSQL at the time.

This commit adds support for virtual generated columns on PostgreSQL 18 and later:

- `t.virtual ... stored: false` now generates virtual generated columns.
- Omitting `stored:` also creates virtual generated columns on PostgreSQL 18+.

On PostgreSQL < 18, specifying `stored: false` or omitting `stored:` raises ArgumentError. consistent with the existing behavior in earlier Rails versions.

### Additional information

Reference: PostgreSQL 18 release notes
https://www.postgresql.org/docs/18/release-18.html

> Allow generated columns to be virtual, and make them the default
> (Peter Eisentraut, Jian He, Richard Guo, Dean Rasheed)
>
> Virtual generated columns generate their values when the columns are read, not written.
> The write behavior can still be specified via the STORED option.

Source:
- https://git.postgresql.org/gitweb/?p=postgresql.git;a=commitdiff;h=83ea6c540
- https://git.postgresql.org/gitweb/?p=postgresql.git;a=commitdiff;h=cdc168ad4
- https://git.postgresql.org/gitweb/?p=postgresql.git;a=commitdiff;h=1e4351af3

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

fatkodima commented on activerecord/lib/active_record/connection_adapters/postgresql/column.rb:
```diff
@@ -26,8 +26,11 @@ def auto_incremented_by_db?
         def virtual?
-          # We assume every generated column is virtual, no matter the concrete type
-          @generated.present?
+          @generated == "s" || @generated == "v"
```
Doesn't the previous implementation still work?

fatkodima commented on activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb:
```diff
@@ -126,16 +127,17 @@ def add_column_options!(sql, options)
-              if options[:stored]
-                sql << " STORED"
-              else
+              if stored != true && database_version < 180_000
```
```suggestion
              if stored != true && database_version < 18_00_00
```

fatkodima commented on activerecord/test/cases/adapters/postgresql/virtual_column_test.rb:
```diff
@@ -12,14 +12,28 @@ class VirtualColumn < ActiveRecord::Base
-        t.virtual :name_octet_length, type: :integer, as: "OCTET_LENGTH(name)", stored: true
-        t.integer :column1
-        t.virtual :column2, type: :integer, as: "column1 + 1", stored: true
+      if @connection.database_version >= 18_00_00
+        @connection.create_table :virtual_columns, force: true do |t|
+          t.string  :name
+          t.virtual :upper_name,  type: :string,  as: "UPPER(name)", stored: true
+          t.virtual :name_length, type: :integer, as: "LENGTH(name)", stored: true
+          t.virtual :name_octet_length, type: :integer, as: "OCTET_LENGTH(name)", stored: true
+          t.integer :column1
+          t.virtual :column2, type: :integer, as: "column1 + 1", stored: true
+          t.virtual :column3, type: :integer, as: "column1 + 2", stored: false
+          t.virtual :column4, type: :integer, as: "column1 + 3"
+        end
+      else
+        @connection.create_table :virtual_columns, force: true do |t|
+          t.string  :name
+          t.virtual :upper_name,  type: :string,  as: "UPPER(name)", stored: true
+          t.virtual :name_length, type: :integer, as: "LENGTH(name)", stored: true
+          t.virtual :name_octet_length, type: :integer, as: "OCTET_LENGTH(name)", stored: true
+          t.integer :column1
+          t.virtual :column2, type: :integer, as: "column1 + 1", stored: true
+        end
```
```suggestion      
      @connection.create_table :virtual_columns, force: true do |t|
        t.string  :name
        t.virtual :upper_name,  type: :string,  as: "UPPER(name)", stored: true
        t.virtual :name_length, type: :integer, as: "LENGTH(name)", stored: true
        t.virtual :name_octet_length, type: :integer, as: "OCTET_LENGTH(name)", stored: true
        t.integer :column1
        t.virtual :column2, type: :integer, as: "column1 + 1", stored: true
        if @connection.database_version >= 18_00_00
          t.virtual :column3, type: :integer, as: "column1 + 2", stored: false
          t.virtual :column4, type: :integer, as: "column1 + 3"
        end
      end
```

fatkodima commented on activerecord/test/cases/adapters/postgresql/virtual_column_test.rb:
```diff
@@ -47,6 +61,7 @@ def test_virtual_column
     def test_stored_column
       column = VirtualColumn.columns_hash["name_length"]
       assert_predicate column, :virtual?
+      assert_predicate column, :virtual_stored? if @connection.database_version >= 18_000
```
```suggestion
      assert_predicate column, :virtual_stored? if @connection.database_version >= 18_00_00
```

kamipo commented on activerecord/lib/active_record/connection_adapters/postgresql/schema_dumper.rb:
```diff
@@ -76,7 +76,7 @@ def prepare_column_options(column)
             if @connection.supports_virtual_columns? && column.virtual?
               spec[:as] = extract_expression_for_virtual_column(column)
-              spec[:stored] = true
+              spec[:stored] = column.virtual_stored?
```
How about matching it with the mysql's output?

```suggestion
              spec[:stored] = "true" if column.virtual_stored?
```

https://github.com/rails/rails/blob/4b7fb1d14b954d2348d2065ff955466479509656/activerecord/lib/active_record/connection_adapters/mysql/schema_dumper.rb#L17-L19

yahonda committed:
Support virtual (not persisted) generated columns on PostgreSQL 18+

PostgreSQL 18 Beta 1 introduces virtual (not persisted) generated columns,
which are now the default unless the STORED option is explicitly specified.

Rails previously supported only stored generated columns on PostgreSQL 12+,
via https://github.com/rails/rails/pull/41856, as only stored generated columns
were supported by PostgreSQL at the time.

This commit adds support for virtual generated columns on PostgreSQL 18 and later:

- `t.virtual ... stored: false` now generates virtual generated columns.
- Omitting `stored:` also creates virtual generated columns on PostgreSQL 18+.
- On PostgreSQL < 18, specifying `stored: false` or omitting `stored:` raises ArgumentError.
  consistent with the existing behavior in earlier Rails versions.

Reference: PostgreSQL 18 release notes
https://www.postgresql.org/docs/18/release-18.html

> Allow generated columns to be virtual, and make them the default
> (Peter Eisentraut, Jian He, Richard Guo, Dean Rasheed)
>
> Virtual generated columns generate their values when the columns are read, not written.
> The write behavior can still be specified via the STORED option.

Source:
- https://git.postgresql.org/gitweb/?p=postgresql.git;a=commitdiff;h=83ea6c540
- https://git.postgresql.org/gitweb/?p=postgresql.git;a=commitdiff;h=cdc168ad4
- https://git.postgresql.org/gitweb/?p=postgresql.git;a=commitdiff;h=1e4351af3

Co-authored-by: fatkodima <fatkodima123@gmail.com>
Co-authored-by: Ryuta Kamizono <kamipo@gmail.com>

yahonda commented:
> We will need to use mariadb:11.4 that is the previous versions of MariaDB lts that should not be affected by https://github.com/rails/rails/issues/53727

Opened https://github.com/rails/buildkite-config/pull/145
